netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java

```diff
@@ -137,37 +135,100 @@ public void creatingMultipleTlsTransportsShouldSucceed() throws Exception {
     }
 
     // Send a single RPC on each transport.
-    final List<SettableFuture> rpcFutures = new ArrayList<SettableFuture>(transports.size());
-    MethodDescriptor<String, String> method = MethodDescriptor.create(MethodType.UNARY,
-            "/testService/test", 10, TimeUnit.SECONDS, StringMarshaller.INSTANCE,
-            StringMarshaller.INSTANCE);
+    final List<Rpc> rpcs = new ArrayList<Rpc>(transports.size());
     for (NettyClientTransport transport : transports) {
-      SettableFuture rpcFuture = SettableFuture.create();
-      rpcFutures.add(rpcFuture);
-      ClientStream stream = transport.newStream(method, new Metadata.Headers(),
-              new TestClientStreamListener(rpcFuture));
-      stream.request(1);
-      stream.writeMessage(messageStream());
-      stream.halfClose();
+      rpcs.add(new Rpc(transport).halfClose());
     }
 
     // Wait for the RPCs to complete.
-    for (SettableFuture rpcFuture : rpcFutures) {
-      rpcFuture.get(10, TimeUnit.SECONDS);
+    for (Rpc rpc : rpcs) {
+      rpc.waitForResponse();
+    }
+  }
+
+  @Test
+  public void bufferedStreamsShouldBeClosedWhenConnectionTerminates() throws Exception {
+    // Only allow a single stream active at a time.
+    startServer(1);
+
+    // Create the protocol negotiator.
+    File clientCert = TestUtils.loadCert("ca.pem");
+    SslContext clientContext = GrpcSslContexts.forClient().trustManager(clientCert).build();
+    ProtocolNegotiator negotiator = ProtocolNegotiators.tls(clientContext, address);
+
+    NettyClientTransport transport = new NettyClientTransport(address, NioSocketChannel.class,
+            group, negotiator, DEFAULT_WINDOW_SIZE, DEFAULT_WINDOW_SIZE);
+    transports.add(transport);
+    transport.start(clientTransportListener);
+
+    // Send a dummy RPC in order to ensure that the updated SETTINGS_MAX_CONCURRENT_STREAMS
+    // has been received by the remote endpoint.
+    new Rpc(transport).halfClose().waitForResponse();
+
+    // Create 3 streams, but don't half-close. The transport will buffer the second and third.
+    Rpc[] rpcs = new Rpc[] { new Rpc(transport), new Rpc(transport), new Rpc(transport) };
+
+    // Wait for the response for the stream that was actually created.
+    rpcs[0].waitForResponse();
+
+    // Now forcibly terminate the connection from the server side.
+    serverListener.transports.get(0).channel().pipeline().firstContext().close();
+
+    // Now wait for both listeners to be closed.
+    for (Rpc rpc : rpcs) {
+      try {
+        rpc.waitForClose();
+        fail("Expected the RPC to fail");
+      } catch (ExecutionException e) {
+        // Expected.
+      }
     }
   }
 
-  private static InputStream messageStream() {
-    return new ByteArrayInputStream(MESSAGE.getBytes());
+  private void startServer(int maxStreamsPerConnection) throws IOException {
+    File serverCert = TestUtils.loadCert("server1.pem");
```

Comment:
Should TLS just be for the test that needs it? It seems like there is a lot of boiler plate here and for the client.